database.cc

```diff
@@ -846,9 +846,28 @@ future<> database::parse_system_tables(distributed<service::storage_proxy>& prox
     }).then([&proxy, &mm, this] {
         return do_parse_schema_tables(proxy, db::schema_tables::VIEWS, [this, &proxy, &mm] (schema_result_value_type &v) {
             return create_views_from_schema_partition(proxy, v.second).then([this, &mm] (std::vector<view_ptr> views) {
-                return parallel_for_each(views.begin(), views.end(), [this, &mm] (auto&& v) {
-                    return this->add_column_family_and_make_directory(v).then([this, &mm, v] {
-                        return maybe_update_legacy_secondary_index_mv_schema(mm.local(), *this, v);
+                return do_with(std::vector<view_ptr>{}, [this, &mm, views = std::move(views)] (std::vector<view_ptr>& views_to_announce) {
+                    return parallel_for_each(views.begin(), views.end(), [this, &mm, &views_to_announce] (auto&& v) {
+                        // TODO: Remove once computed columns are guaranteed to be featured in the whole cluster.
+                        // we fix here the schema in place in oreder to avoid races (write commands comming from other coordinators).
+                        view_ptr fixed_v = maybe_fix_legacy_secondary_index_mv_schema(*this, v, nullptr, preserve_version::yes);
+                        // TODO: Remove once computed columns are guaranteed to be featured in the whole cluster.
+                        view_ptr v_to_add = v;
+                        if (bool(fixed_v)) {
+                            v_to_add = fixed_v;
+                            // now really fix the view by creating a new and updated version and registering it for announcement.
+                            views_to_announce.emplace_back(maybe_fix_legacy_secondary_index_mv_schema(*this, v, nullptr, preserve_version::no));
+                        }
+                        return this->add_column_family_and_make_directory(v_to_add);
+                    }).then([&views_to_announce, &mm, this] {
+                        if (!views_to_announce.empty()) {
+                            _computed_columns_reg = _feat.cluster_supports_computed_columns().when_enabled([&mm, views_to_announce = std::move(views_to_announce)] {
+                                (void)parallel_for_each(views_to_announce.begin(), views_to_announce.end(), [&mm] (const view_ptr& fixed_v) {
```

Comment:
I realized that I put my comments in some obscure commit outside of the tree, so I'll copy everything here:


This orphaned future should be marked with //FIXME: future discarded.. I suppose it would also be possible to implement a sensible future-based version of when_enabled, but that's out of scope for this patch.